File: SBiswas/Problems/Milestone1/Problem20.cpp
/* 
    Problem 20: Range Sum of BST
*/

#include<iostream>
#include<vector>
#include<limits.h>

using namespace std;


//  Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
 
class Solution {
public:

    void helper(TreeNode* root, int& low, int& high, int& sum)
    {
        if(!root)
        {
            return;
        }
        
        helper(root->left, low, high, sum);
        if(root->val<=high && root->val>=low)
        {
            sum += root->val;
        }
        helper(root->right, low, high, sum);

    }

    int rangeSumBST(TreeNode* root, int low, int high) {
        int ans =0;

        helper(root, low, high, ans);

        return ans;
    }
};

File: Hritwik/Problems/Milestone1/Problem20.py
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
    
class Solution:       
    def rangeSumBST(self, root, low: int, high: int) -> int:
        if root is None:
            return 0
        queue = deque([root])
        sum = 0
        while queue:
            for _ in range(len(queue)):
                node = queue.popleft()
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                if node.val>=low and node.val<=high:
                    sum+=node.val
        return sum

solution = Solution()

# Create a binary tree to traverse through it in an inorder traversal manner
#
#         Root   2
#              /    \
#             2      5  
#                   /  \
#                  5    7
#
root = TreeNode(2)
root.left = TreeNode(2)
root.right = TreeNode(5)
root.right.left = TreeNode(5)
root.right.right = TreeNode(7)

low, high = 3, 6
response = solution.rangeSumBST(root, low, high)
print(f'The sum of values of all nodes with a value in the inclusive range [low, high] is {response}')
